src/main/scala/org/opennetworkinsight/proxy/ProxySuspiciousConnectsModel.scala

```diff
@@ -0,0 +1,192 @@
+package org.opennetworkinsight.proxy
+
+import org.apache.spark.SparkContext
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+import org.opennetworkinsight.OniLDACWrapper
+import org.opennetworkinsight.OniLDACWrapper.{OniLDACInput, OniLDACOutput}
+import org.opennetworkinsight.SuspiciousConnectsArgumentParser.SuspiciousConnectsConfig
+import org.opennetworkinsight.utilities._
+import org.slf4j.Logger
+import org.opennetworkinsight.proxy.ProxySchema._
+
+/**
+  * Encapsulation of a proxy suspicious connections model.
+  *
+  * @param topicCount         Number of "topics" used to cluster IPs and proxy "words" in the topic modelling analysis.
+  * @param ipToTopicMIx       Maps each IP to a vector measuring Prob[ topic | this IP] for each topic.
+  * @param wordToPerTopicProb Maps each word to a vector measuring Prob[word | topic] for each topic.
+  * @param timeCuts           Decile cutoffs for time-of-day in seconds.
+  * @param entropyCuts        Quintile cutoffs for measurement of full URI string entropy.
+  * @param agentCuts          Quintiile cutoffs for frequency of user agent.
+  */
+class ProxySuspiciousConnectsModel(topicCount: Int,
+                                   ipToTopicMIx: Map[String, Array[Double]],
+                                   wordToPerTopicProb: Map[String, Array[Double]],
+                                   timeCuts: Array[Double],
+                                   entropyCuts: Array[Double],
+                                   agentCuts: Array[Double]) {
+
+
+  /**
+    * Calculate suspicious connection scores for an incoming dataframe using this proxy suspicious connects model.
+    *
+    * @param sc        Spark context.
+    * @param dataFrame Dataframe with columns Host, Time, ReqMethod, FullURI, ResponseContentType, UserAgent, RespCode
+    *                  (as defined in ProxySchema object).
+    * @return Dataframe with Score column added.
+    */
+  def score(sc: SparkContext, dataFrame: DataFrame): DataFrame = {
+
+    val topDomains: Broadcast[Set[String]] = sc.broadcast(TopDomains.TOP_DOMAINS)
+
+    val agentToCount: Map[String, Long] =
+      dataFrame.select(UserAgent).rdd.map({ case Row(ua: String) => (ua, 1L) }).reduceByKey(_ + _).collect().toMap
+
+    val agentToCountBC = sc.broadcast(agentToCount)
+
+    val udfWordCreation =
+      ProxyWordCreation.udfWordCreation(topDomains, agentToCountBC, timeCuts, entropyCuts, agentCuts)
```

Comment:
It's great we are not re-calculating timeCuts, entropyCuts and agentCuts, those are some of the most expensive actions in our process. With them happening only once we are going to reduce the time significantly but I'm thinking and asking myself - since the other day- if we really need to re-calculate or re-create data with words. Essentially it should be the same data set used during the training and the scoring that's why we are carrying the rawData data frame. What if we cache the entire data frame with words instead of just returning the cuts? And that way we skip the word creation.

I know this could not make sense in case we want to break the logic - training and scoring - in the future, but for now it looks to me they are interdependent so we could take advantage of that.

There are two possible ways: 
1. registerTempTable
2. persist